File: Assets/Bundles/Shared/Resources/TS/GameObjectBridge.lua
-- Compiled with unity-ts v2.1.0-75
local TS = require("Shared/include/RuntimeLib")
local RunUtil = require("Shared/TS/Util/RunUtil").RunUtil
-- * Wrapper around `Object` functionality.
local GameObjectBridge
do
	GameObjectBridge = setmetatable({}, {
		__tostring = function()
			return "GameObjectBridge"
		end,
	})
	GameObjectBridge.__index = GameObjectBridge
	function GameObjectBridge.new(...)
		local self = setmetatable({}, GameObjectBridge)
		return self:constructor(...) or self
	end
	function GameObjectBridge:constructor()
	end
	function GameObjectBridge:Instantiate(original, tag)
		-- * Fire `GameObjectInstantiated` event if tagged.
		local go = Object:Instantiate(original)
		if tag then
			GameObjectBridge:FireTaggedGameObjectInstantiatedSignal(go, tag)
		end
		return go
	end
	function GameObjectBridge:InstantiateAt(original, position, rotation, tag)
		-- * Fire `GameObjectInstantiated` event if tagged.
		local go = Object:Instantiate(original, position, rotation)
		if tag then
			GameObjectBridge:FireTaggedGameObjectInstantiatedSignal(go, tag)
		end
		return go
	end
	function GameObjectBridge:InstantiateIn(original, parent, tag)
		-- * Fire `GameObjectInstantiated` event if tagged.
		local go = Object:Instantiate(original, parent)
		if tag then
			GameObjectBridge:FireTaggedGameObjectInstantiatedSignal(go, tag)
		end
		return go
	end
	function GameObjectBridge:Destroy(gameObject, delay)
		if delay ~= nil then
			Object:Destroy(gameObject, delay)
		else
			Object:Destroy(gameObject)
		end
	end
	function GameObjectBridge:FireTaggedGameObjectInstantiatedSignal(gameObject, tag)
		if RunUtil:IsServer() then
			local _promise = TS.Promise.new(function(resolve)
				resolve(require("Server/TS/ServerSignals"))
			end)
			local _arg0 = function(serverSignalsRef)
				local tagAddedSignal = serverSignalsRef.ServerSignals.CollectionManagerTagAdded
				tagAddedSignal:Fire({
					go = gameObject,
					tag = tag,
				})
			end
			_promise:andThen(_arg0)
		elseif RunUtil:IsClient() then
			local _promise = TS.Promise.new(function(resolve)
				resolve(require("Client/TS/ClientSignals"))
			end)
			local _arg0 = function(clientSignalsRef)
				local tagAddedSignal = clientSignalsRef.ClientSignals.CollectionManagerTagAdded
				tagAddedSignal:Fire({
					go = gameObject,
					tag = tag,
				})
			end
			_promise:andThen(_arg0)
		end
	end
end
return {
	GameObjectBridge = GameObjectBridge,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Shared/GameObjectBridge.ts
import { CollectionTag } from "./Util/CollectionTag";
import { RunUtil } from "./Util/RunUtil";

/** Wrapper around `Object` functionality. */
export class GameObjectUtil {
	/**
	 * Instantiate an Object.
	 * @param original The object being instantiated.
	 * @param tag An optional tag to register `GameObject` with `CollectionManager`.
	 * @returns The instantiated GameObject.
	 */
	public static Instantiate(original: Object, tag?: CollectionTag): GameObject {
		/** Fire `GameObjectInstantiated` event if tagged. */
		const go = Object.Instantiate(original) as GameObject;
		if (tag) {
			GameObjectUtil.FireTaggedGameObjectInstantiatedSignal(go, tag);
		}
		return go;
	}

	/**
	 * Instantiate an Object positionally.
	 * @param original The object being instantiated.
	 * @param position The position to instantiate object at.
	 * @param rotation The rotation with instantiate object with.
	 * @param tag An optional tag to register `GameObject` with `CollectionManager`.
	 * @returns The instantiated GameObject.
	 */
	public static InstantiateAt(
		original: Object,
		position: Vector3,
		rotation: Quaternion,
		tag?: CollectionTag,
	): GameObject {
		/** Fire `GameObjectInstantiated` event if tagged. */
		const go = Object.Instantiate(original, position, rotation) as GameObject;
		if (tag) {
			GameObjectUtil.FireTaggedGameObjectInstantiatedSignal(go, tag);
		}
		return go;
	}

	/**
	 * Instantiate an Object as a child of `parent`.
	 * @param original The object being instantiated.
	 * @param parent The object to instantiate as a child of.
	 * @param tag An optional tag to register `GameObject` with `CollectionManager`.
	 * @returns The instantiated GameObject.
	 */
	public static InstantiateIn(original: Object, parent: Transform, tag?: CollectionTag): GameObject {
		/** Fire `GameObjectInstantiated` event if tagged. */
		const go = Object.Instantiate(original, parent) as GameObject;
		if (tag) {
			GameObjectUtil.FireTaggedGameObjectInstantiatedSignal(go, tag);
		}
		return go;
	}

	/**
	 * Destroys a GameObject.
	 * @param gameObject GameObject to destroy.
	 * @param delay The optional amount of time to delay before destroying the object.
	 */
	public static Destroy(gameObject: GameObject, delay?: number): void {
		if (delay !== undefined) {
			Object.Destroy(gameObject, delay);
		} else {
			Object.Destroy(gameObject);
		}
	}

	/** Fire relevant `GameObjectInstantiated` event based on runtime. */
	private static FireTaggedGameObjectInstantiatedSignal(gameObject: GameObject, tag: CollectionTag): void {
		if (RunUtil.IsServer()) {
			import("Server/ServerSignals").then((serverSignalsRef) => {
				const tagAddedSignal = serverSignalsRef.ServerSignals.CollectionManagerTagAdded;
				tagAddedSignal.Fire({ go: gameObject, tag: tag });
			});
		} else if (RunUtil.IsClient()) {
			import("Client/ClientSignals").then((clientSignalsRef) => {
				const tagAddedSignal = clientSignalsRef.ClientSignals.CollectionManagerTagAdded;
				tagAddedSignal.Fire({ go: gameObject, tag: tag });
			});
		}
	}
}
